### PR TITLE
[mycpp] add maxsplits flavor for Str::split()

### PR DESCRIPTION
**Pull request by melvinw**
Some yet-to-be-translated code uses this.

### CODE REVIEW

**Review by andychu — COMMENTED** [review]
Thanks, I started using this handy second param a lot!

But one thing I forgot is that we have `mylib.split_once()` because it's more easily statically typed, and allocates less.

That is for the case when you have `KEY=value` or `KEY=value=z`, but you don't care what's after the first `=`

So some stuff might be worth rewriting to mylib.split_once(), although I won't be surprised if we need this too

File: mycpp/gc_str_test.cc [hunk 1]
```diff
@@ -997,6 +997,20 @@ TEST test_str_split() {
...
+    PRINT_LIST(result);
+    ASSERT(len(result) == 7);
+    ASSERT(are_equal(result->index_(0), StrFromC("a")));
+
+    result = (StrFromC("a,b,c,d,e,f,g"))->split(StrFromC(","), 3);
+    PRINT_LIST(result);
+    ASSERT(len(result) == 3);
+    ASSERT(are_equal(result->index_(0), StrFromC("a")));
+    ASSERT(are_equal(result->index_(1), StrFromC("b")));
+    ASSERT(are_equal(result->index_(2), StrFromC("c")));
```
[the remark this answers is not in the document]
Comment (melvinw) [reply mycpp/gc_str_test.cc:1012]: Oh nice catch! Pushed a fix

File: mycpp/gc_str.cc [hunk 2]
```diff
@@ -442,7 +442,7 @@ List<Str*>* Str::split(Str* sep) {
   int pos = 0;
   int end = n;
 
-  while (true) {
+  while (len(result) < max_split) {
```
Comment (andychu) [inline mycpp/gc_str.cc:445]: Hm one thing that is a bit suspicious is that there can be 2 append() to the list in a single iteration in some cases, i.e. with kEmptyString

Not sure if that actually causes a bug or not

I think this function was structured somewhat oddly, so I might poke at it a little more

**Review by andychu — COMMENTED** [review]
Cool thanks for adding the tests

**Comment by andychu** [comment]
Hm actually when I grep I only see `max_splits == 1`, which makes me suspect that mylib.split_once() could suffice

Let me check it out

```
$ metrics/source-code.sh osh-files  |xargs grep 'split('
bin/oil.py:    command, arg = msg.split(' ', 1)
core/completion.py:    dirname, basename = os_path.split(to_complete)
core/completion.py:    path_dirs = val.s.split(':')
core/comp_ui.py:  last_line = display_str.split('\n')[-1]
core/main_loop.py:          command, arg = blob.split(' ', 1)
core/pyutil.py:  py_version = version_line.split()[0]
core/state.py:      path_list = val.s.split(':')
core/state.py:    lookup = shellopts.split(':')
core/state.py:        names = [n for n in shellopts.split(':') if n != opt_name]
osh/builtin_comp.py:        prev, cur = cur.split('=', 1)
osh/builtin_pure.py:          path = name.split('/')
osh/split.py:    Split used by word evaluation.  Also used by the explicit @split() functino.
osh/split.py:    """split() function exposed to users"""
osh/split.py:        e_die('split() passed arg of invalid type %r', s.__class__.__name__)
osh/tdop.py:    namespace = f.__module__.split('.')[-1]
```

**Comment by andychu** [comment]
OK I cloned this and the tests look good, and I wasn't able to tickle the apparent issue where there are multiple string append per loop, but only one maxsplits check. 

The thing that I find odd is that there were 3 different places to append(), and now there are 4.  Maybe I will massage it since there are good tests

----

Also I remembered why we're using `mylib.split_once()` -- because unpacking like

```
x, y = foo.split(' ', 1)
```

does NOT work, because the number of items returned isn't static (it depends on a parameter).

So we will have to rewrite them all with mylib.split_once() anyway.  That is technically better since it saves a temporary heap allocation of a `List` and `Slab`

Eventually we will "hoist" split() to the user level, so it's good to have an implementation

**Pull request merged**